File: application/puzzles.py
from __future__ import annotations

from dataclasses import dataclass
import random
from typing import Dict, Optional, Tuple

from domain.types import Digit

Grid = Tuple[Tuple[Optional[Digit], ...], ...]


@dataclass(frozen=True)
class Puzzle:
    """A Sudoku puzzle definition.

    - grid uses None for empty cells and 1..9 for givens.
    - difficulty is a simple label for now; refine later.
    """

    puzzle_id: str
    difficulty: str
    grid: Grid


# Seed puzzles (None for empty). These are valid Sudoku puzzles used as templates.
_EASY_GRID_01: Grid = (
    (5, 3, None, None, 7, None, None, None, None),
    (6, None, None, 1, 9, 5, None, None, None),
    (None, 9, 8, None, None, None, None, 6, None),
    (8, None, None, None, 6, None, None, None, 3),
    (4, None, None, 8, None, 3, None, None, 1),
    (7, None, None, None, 2, None, None, None, 6),
    (None, 6, None, None, None, None, 2, 8, None),
    (None, None, None, 4, 1, 9, None, None, 5),
    (None, None, None, None, 8, None, None, 7, 9),
)

# A medium template (classic example puzzle)
_MEDIUM_GRID_01: Grid = (
    (None, None, 3, None, 2, None, 6, None, None),
    (9, None, None, 3, None, 5, None, None, 1),
    (None, None, 1, 8, None, 6, 4, None, None),
    (None, None, 8, 1, None, 2, 9, None, None),
    (7, None, None, None, None, None, None, None, 8),
    (None, None, 6, 7, None, 8, 2, None, None),
    (None, None, 2, 6, None, 9, 5, None, None),
    (8, None, None, 2, None, 3, None, None, 9),
    (None, None, 5, None, 1, None, 3, None, None),
)

# A harder template (fewer givens)
_HARD_GRID_01: Grid = (
    (None, None, None, None, None, None, None, 1, 2),
    (None, None, None, None, 3, 5, None, None, None),
    (None, None, None, 7, None, None, 3, None, None),
    (None, 3, None, None, None, None, None, None, None),
    (1, None, None, None, None, None, None, None, 6),
    (None, None, None, None, None, None, None, 7, None),
    (None, None, 5, None, None, 8, None, None, None),
    (None, None, None, 2, 9, None, None, None, None),
    (7, 2, None, None, None, None, None, None, None),
)


_EASY_SEEDS: Tuple[Grid, ...] = (_EASY_GRID_01,)
_MEDIUM_SEEDS: Tuple[Grid, ...] = (_MEDIUM_GRID_01,)
_HARD_SEEDS: Tuple[Grid, ...] = (_HARD_GRID_01,)


_DIFFICULTY_SEEDS: Dict[str, Tuple[Grid, ...]] = {
    "easy": _EASY_SEEDS,
    "medium": _MEDIUM_SEEDS,
    "hard": _HARD_SEEDS,
}


PUZZLES: Dict[str, Puzzle] = {
    "starter": Puzzle(puzzle_id="starter", difficulty="easy", grid=_EASY_GRID_01),
}


def get_puzzle(puzzle_id: str = "starter") -> Puzzle:
    """Fetch a puzzle by id.

    Raises:
        KeyError: if the puzzle_id is not known.
    """

    return PUZZLES[puzzle_id]


def list_puzzles() -> Tuple[Puzzle, ...]:
    """Return all available puzzles as an immutable tuple."""

    # Build a tuple via comprehension so the element type is unambiguous
    return tuple(p for p in PUZZLES.values())



def _randomize_seed(seed: Grid, rng: random.Random) -> Grid:
    """Randomize a seed grid while preserving Sudoku validity.

    Uses only structure-preserving transforms:
    - digit relabeling (a permutation of 1..9)
    - row permutations within each 3-row band, plus band permutations
    - column permutations within each 3-col stack, plus stack permutations

    The pattern of filled cells remains unchanged.
    """

    # 1) Digit relabeling: create a bijection 1..9 -> 1..9
    perm = list(range(1, 10))
    rng.shuffle(perm)
    digit_map = {d: perm[d - 1] for d in range(1, 10)}

    # 2) Row ordering: shuffle bands and rows within each band
    bands = [0, 1, 2]
    rng.shuffle(bands)
    row_order = []
    for b in bands:
        rows = [0, 1, 2]
        rng.shuffle(rows)
        row_order.extend([b * 3 + r for r in rows])

    # 3) Column ordering: shuffle stacks and cols within each stack
    stacks = [0, 1, 2]
    rng.shuffle(stacks)
    col_order = []
    for s in stacks:
        cols = [0, 1, 2]
        rng.shuffle(cols)
        col_order.extend([s * 3 + c for c in cols])

    # 4) Apply permutations
    new_rows = []
    for r in row_order:
        row = []
        for c in col_order:
            v = seed[r][c]
            if v is None:
                row.append(None)
            else:
                row.append(digit_map[v])
        new_rows.append(tuple(row))

    return tuple(new_rows)

# --- Randomized starter generator


def generate_puzzle(difficulty: str = "easy", rng: random.Random | None = None) -> Puzzle:
    """Generate a puzzle for the requested difficulty.

    This generates a full solved grid and then masks cells according to difficulty.

    Args:
        difficulty: "easy" | "medium" | "hard" (case-insensitive).
        rng: Optional random generator (injectable for deterministic tests).

    Returns:
        A new Puzzle instance.

    Raises:
        KeyError: if the difficulty is unknown.
    """

    if rng is None:
        rng = random.Random()

    diff = (difficulty or "easy").strip().lower()
    if diff not in ("easy", "medium", "hard"):
        raise KeyError(diff)

    solution = _generate_full_solution(rng)
    grid = _mask_solution(solution, diff, rng)

    puzzle_id = diff + "_gen_" + format(rng.getrandbits(32), "08x")
    return Puzzle(puzzle_id=puzzle_id, difficulty=diff, grid=grid)


def generate_random_starter(rng: random.Random | None = None) -> Puzzle:
    """Backward-compatible alias for an easy randomized starter."""

    return generate_puzzle("easy", rng=rng)


# --- Local legality checker and full-solution generator


def _is_legal(grid: list[list[Optional[Digit]]], r: int, c: int, d: Digit) -> bool:
    """Return True iff placing digit d at (r, c) does not violate Sudoku constraints."""

    # Row
    for cc in range(9):
        if grid[r][cc] == d:
            return False

    # Column
    for rr in range(9):
        if grid[rr][c] == d:
            return False

    # Box
    br = (r // 3) * 3
    bc = (c // 3) * 3
    for rr in range(br, br + 3):
        for cc in range(bc, bc + 3):
            if grid[rr][cc] == d:
                return False

    return True


def _generate_full_solution(rng: random.Random) -> Grid:
    """Generate a complete solved Sudoku grid using randomized backtracking."""

    work: list[list[Optional[Digit]]] = [[None] * 9 for _ in range(9)]
    digits: list[Digit] = [1, 2, 3, 4, 5, 6, 7, 8, 9]

    def find_empty() -> tuple[int, int] | None:
        for rr in range(9):
            for cc in range(9):
                if work[rr][cc] is None:
                    return rr, cc
        return None

    def fill() -> bool:
        spot = find_empty()
        if spot is None:
            return True

        rr, cc = spot
        candidates = digits[:]
        rng.shuffle(candidates)

        for val in candidates:
            if _is_legal(work, rr, cc, val):
                work[rr][cc] = val
                if fill():
                    return True
                work[rr][cc] = None

        return False

    if not fill():
        raise RuntimeError("Unable to generate a solved Sudoku grid")

    return tuple(tuple(work[r][c] for c in range(9)) for r in range(9))


# --- Phase 2: Masking a solved grid into a puzzle

# Target givens per difficulty (initial heuristic; refine later)
_TARGET_GIVENS: Dict[str, int] = {
    "easy": 40,
    "medium": 32,
    "hard": 26,
}


def _mask_solution(solution: Grid, difficulty: str, rng: random.Random) -> Grid:
    """Mask (remove) values from a solved grid to form a puzzle.

    Guarantees at least one solution (the provided solution). Uniqueness is NOT
    enforced yet.
    """

    diff = (difficulty or "easy").strip().lower()
    target = _TARGET_GIVENS.get(diff, _TARGET_GIVENS["easy"])

    work: list[list[Optional[Digit]]] = [list(row) for row in solution]

    # Rotational symmetry (180°): remove (r, c) together with (8-r, 8-c).
    # We iterate only over half the grid (including the center cell) to avoid duplicates.
    coords = [(r, c) for r in range(9) for c in range(9) if (r < 4) or (r == 4 and c <= 4)]
    rng.shuffle(coords)

    givens = 81
    for r, c in coords:
        if givens <= target:
            break

        r2, c2 = 8 - r, 8 - c

        # Determine how many givens would be removed by this symmetric step.
        # Center cell (4,4) maps to itself.
        if r == r2 and c == c2:
            removal = 1
        else:
            removal = 2

        # Do not overshoot below the target; preserve symmetry.
        if givens - removal < target:
            continue

        # Remove the symmetric pair if present.
        old1 = work[r][c]
        old2 = work[r2][c2]

        if old1 is None and old2 is None:
            continue

        work[r][c] = None
        work[r2][c2] = None
        givens -= removal

    return tuple(tuple(work[r][c] for c in range(9)) for r in range(9))

File: flutter_app/lib/application/puzzles.dart
import 'dart:math';

import 'package:flutter_app/domain/types.dart';

class Puzzle {
  final String puzzleId;
  final String difficulty;
  final Grid grid;

  const Puzzle({
    required this.puzzleId,
    required this.difficulty,
    required this.grid,
  });
}

const Grid _easyGrid01 = [
  [5, 3, null, null, 7, null, null, null, null],
  [6, null, null, 1, 9, 5, null, null, null],
  [null, 9, 8, null, null, null, null, 6, null],
  [8, null, null, null, 6, null, null, null, 3],
  [4, null, null, 8, null, 3, null, null, 1],
  [7, null, null, null, 2, null, null, null, 6],
  [null, 6, null, null, null, null, 2, 8, null],
  [null, null, null, 4, 1, 9, null, null, 5],
  [null, null, null, null, 8, null, null, 7, 9],
];

const Grid _mediumGrid01 = [
  [null, null, 3, null, 2, null, 6, null, null],
  [9, null, null, 3, null, 5, null, null, 1],
  [null, null, 1, 8, null, 6, 4, null, null],
  [null, null, 8, 1, null, 2, 9, null, null],
  [7, null, null, null, null, null, null, null, 8],
  [null, null, 6, 7, null, 8, 2, null, null],
  [null, null, 2, 6, null, 9, 5, null, null],
  [8, null, null, 2, null, 3, null, null, 9],
  [null, null, 5, null, 1, null, 3, null, null],
];

const Grid _hardGrid01 = [
  [null, null, null, null, null, null, null, 1, 2],
  [null, null, null, null, 3, 5, null, null, null],
  [null, null, null, 7, null, null, 3, null, null],
  [null, 3, null, null, null, null, null, null, null],
  [1, null, null, null, null, null, null, null, 6],
  [null, null, null, null, null, null, null, 7, null],
  [null, null, 5, null, null, 8, null, null, null],
  [null, null, null, 2, 9, null, null, null, null],
  [7, 2, null, null, null, null, null, null, null],
];

const Map<String, List<Grid>> _difficultySeeds = {
  'easy': [_easyGrid01],
  'medium': [_mediumGrid01],
  'hard': [_hardGrid01],
};

const Map<String, int> _targetGivens = {
  'easy': 40,
  'medium': 32,
  'hard': 26,
};

final Map<String, Puzzle> puzzles = {
  'starter': Puzzle(puzzleId: 'starter', difficulty: 'easy', grid: _easyGrid01),
};

Puzzle getPuzzle(String puzzleId) {
  final puzzle = puzzles[puzzleId];
  if (puzzle == null) {
    throw ArgumentError('Unknown puzzle id: $puzzleId');
  }
  return puzzle;
}

List<Puzzle> listPuzzles() => puzzles.values.toList(growable: false);

Puzzle generatePuzzle(String difficulty, {Random? rng}) {
  final random = rng ?? Random();
  final diff = difficulty.trim().toLowerCase();
  if (!_difficultySeeds.containsKey(diff)) {
    throw ArgumentError('Unknown difficulty: $difficulty');
  }
  final solution = _generateFullSolution(random);
  final grid = _maskSolution(solution, diff, random);
  final puzzleId = '${diff}_gen_${random.nextInt(0xffffffff).toRadixString(16).padLeft(8, '0')}';
  return Puzzle(puzzleId: puzzleId, difficulty: diff, grid: grid);
}

Puzzle generateRandomStarter({Random? rng}) {
  return generatePuzzle('easy', rng: rng);
}

Grid randomizeSeed(Grid seed, Random rng) {
  final perm = List<int>.generate(9, (index) => index + 1)..shuffle(rng);
  final digitMap = {for (var d = 1; d <= 9; d += 1) d: perm[d - 1]};

  final bands = [0, 1, 2]..shuffle(rng);
  final rowOrder = <int>[];
  for (final band in bands) {
    final rows = [0, 1, 2]..shuffle(rng);
    rowOrder.addAll(rows.map((r) => band * 3 + r));
  }

  final stacks = [0, 1, 2]..shuffle(rng);
  final colOrder = <int>[];
  for (final stack in stacks) {
    final cols = [0, 1, 2]..shuffle(rng);
    colOrder.addAll(cols.map((c) => stack * 3 + c));
  }

  final newRows = <List<Digit?>>[];
  for (final r in rowOrder) {
    final row = <Digit?>[];
    for (final c in colOrder) {
      final value = seed[r][c];
      row.add(value == null ? null : digitMap[value]);
    }
    newRows.add(row);
  }
  return newRows;
}

Grid _generateFullSolution(Random rng) {
  final work = List<List<Digit?>>.generate(
    9,
    (_) => List<Digit?>.filled(9, null, growable: false),
    growable: false,
  );
  final digits = List<int>.generate(9, (index) => index + 1);

  List<int>? findEmpty() {
    for (var r = 0; r < 9; r += 1) {
      for (var c = 0; c < 9; c += 1) {
        if (work[r][c] == null) {
          return [r, c];
        }
      }
    }
    return null;
  }

  bool fill() {
    final spot = findEmpty();
    if (spot == null) {
      return true;
    }
    final rr = spot[0];
    final cc = spot[1];
    final candidates = [...digits]..shuffle(rng);
    for (final val in candidates) {
      if (_isLegal(work, rr, cc, val)) {
        work[rr][cc] = val;
        if (fill()) {
          return true;
        }
        work[rr][cc] = null;
      }
    }
    return false;
  }

  if (!fill()) {
    throw StateError('Unable to generate a solved Sudoku grid');
  }

  return work
      .map((row) => row.map((cell) => cell).toList(growable: false))
      .toList(growable: false);
}

Grid _maskSolution(Grid solution, String difficulty, Random rng) {
  final diff = difficulty.trim().toLowerCase();
  final target = _targetGivens[diff] ?? _targetGivens['easy']!;
  final work = solution.map((row) => row.toList()).toList();
  final coords = <List<int>>[];
  for (var r = 0; r < 9; r += 1) {
    for (var c = 0; c < 9; c += 1) {
      if (r < 4 || (r == 4 && c <= 4)) {
        coords.add([r, c]);
      }
    }
  }
  coords.shuffle(rng);

  var givens = 81;
  for (final coord in coords) {
    if (givens <= target) {
      break;
    }
    final r = coord[0];
    final c = coord[1];
    final r2 = 8 - r;
    final c2 = 8 - c;
    final removal = (r == r2 && c == c2) ? 1 : 2;
    if (givens - removal < target) {
      continue;
    }
    if (work[r][c] == null && work[r2][c2] == null) {
      continue;
    }
    work[r][c] = null;
    work[r2][c2] = null;
    givens -= removal;
  }
  return work
      .map((row) => row.map((cell) => cell).toList(growable: false))
      .toList(growable: false);
}

bool _isLegal(List<List<Digit?>> grid, int r, int c, Digit digit) {
  for (var cc = 0; cc < 9; cc += 1) {
    if (grid[r][cc] == digit) {
      return false;
    }
  }
  for (var rr = 0; rr < 9; rr += 1) {
    if (grid[rr][c] == digit) {
      return false;
    }
  }
  final br = (r ~/ 3) * 3;
  final bc = (c ~/ 3) * 3;
  for (var rr = br; rr < br + 3; rr += 1) {
    for (var cc = bc; cc < bc + 3; cc += 1) {
      if (grid[rr][cc] == digit) {
        return false;
      }
    }
  }
  return true;
}
